### PR TITLE
[SWT-0006] Return the thrown error from `#expect(throws:)` and `#require(throws:)`.

### PR DESCRIPTION
**Pull request by grynspan**
This PR changes the signatures of the various `throws:` overloads of `#expect` and `#require` so that on success they return the error that was thrown rather than `Void`. This then allows more ergonomic inspection of the error's properties:

```swift
let error = try #require(throws: MyError.self) {
  try f()
}
#expect(error.hasWidget)
#expect(error.userName == "John Smith")
```

For more information, see [the proposal document](https://github.com/swiftlang/swift-testing/blob/jgrynspan/return-errors-from-expect-throws/Documentation/Proposals/0006-return-errors-from-expect-throws.md).

Resolves rdar://138235250.

<details>
<summary>Further PR Details</summary>
It is not possible to overload a macro or function solely by return type without the compiler reporting `Ambiguous use of 'f()'`, so we are not able to stage this change in using `@_spi(Experimental)` without breaking test code that already imports our SPI.

This change is potentially source-breaking for tests that inadvertently forward the result of these macro invocations to an enclosing scope. For example, the compiler will start emitting a warning here:

```swift
func bar(_ pfoo: UnsafePointer<Foo>) throws { ... }

withUnsafePointer(to: foo) { pfoo in // ⚠️ Result of call to 'withUnsafePointer(to:_:)' is unused
  #expect(throws: BadFooError.self) {
    try bar(pfoo)
  }
}
```

This warning can be suppressed by assigning the result of `#expect` (or of `withUnsafePointer(to:_:)`) to `_`:

```swift
func bar(_ pfoo: UnsafePointer<Foo>) throws { ... }

withUnsafePointer(to: foo) { pfoo in
  _ = #expect(throws: BadFooError.self) {
    try bar(pfoo)
  }
}
```

Because `#expect` and `#require` are macros, they cannot be referenced by name like functions, so you cannot assign them to variables (and then run into trouble with the types of those variables.)

Finally, this change deprecates the variants of `#expect` and `#require` that take two closures:

```swift
#expect { // ⚠️ 'expect(_:sourceLocation:performing:throws:)' is deprecated: Examine the result of '#expect(throws:)' instead.
  ...
} throws: { error in
  guard let error = error as? FoodTruckError else {
    return false
  }
  return error.napkinCount == 0
}
```

These variants are no longer needed because you can simply examine the result of the other variants:

```swift
let error = #expect(throws: FoodTruckError.self) {
  ...
}
#expect(error?.napkinCount == 0)
```
</details>

### Checklist:

- [x] Code and documentation should follow the style of the [Style Guide](https://github.com/apple/swift-testing/blob/main/Documentation/StyleGuide.md).
- [x] If public symbols are renamed or modified, DocC references should be updated.

### CODE REVIEW

**Comment by grynspan** [comment]
@swift-ci test

**Comment by grynspan** [comment]
@swift-ci test

**Comment by grynspan** [comment]
@swift-ci test

**Comment by grynspan** [comment]
@swift-ci test

**Comment by grynspan** [comment]
@swift-ci test

**Comment by dabrahams** [comment]
Nice!

File: Documentation/Proposals/NNNN-return-errors-from-expect-throws.md [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+```diff
+--- a/Sources/Testing/Expectations/Expectation+Macro.swift
++++ b/Sources/Testing/Expectations/Expectation+Macro.swift
++@discardableResult
+ @freestanding(expression) public macro expect<E, R>(
+   throws errorType: E.Type,
+   _ comment: @autoclosure () -> Comment? = nil,
+   sourceLocation: SourceLocation = #_sourceLocation,
+   performing expression: () async throws -> R
+-) = #externalMacro(module: "TestingMacros", type: "ExpectMacro") where E: Error
```
Comment (grynspan) [inline Documentation/Proposals/NNNN-return-errors-from-expect-throws.md:78]: Need to strip the external macros from the faux diff here.

File: Sources/TestingMacros/Support/Additions/MacroExpansionContextAdditions.swift [hunk 2]
```diff
@@ -80,30 +80,60 @@ extension MacroExpansionContext {
...
+  /// lexical context.
+  ///
+  /// The value of this property is `true` if the current lexical context
+  /// contains a node with the `@_semantics("testing.macros.nowarnings")`
+  /// attribute applied to it.
+  ///
+  /// - Warning: This functionality is not part of the public interface of the
+  ///   testing library. It may be modified or removed in a future update.
+  var areWarningsSuppressed: Bool {
+    for lexicalContext in self.lexicalContext {
```
Comment (grynspan) [inline Sources/TestingMacros/Support/Additions/MacroExpansionContextAdditions.swift:93]: Disable this in release builds so folks don't try to use it!

File: Sources/Testing/Support/Additions/ResultAdditions.swift [hunk 3]
```diff
@@ -37,10 +37,33 @@ extension Result {
...
+  /// special handling for a `nil` result so that `Never.self` (which can't be
+  /// instantiated) can be used as the error type in the macro call.
+  ///
+  /// If the value really is `nil` (i.e. we're dealing with `Never`), the
+  /// testing library throws an error representing an issue of kind
+  /// ``Issue/Kind-swift.enum/apiMisused``.
+  ///
+  /// - Warning: This function is used to implement the `#expect()` and
+  ///   `#require()` macros. Do not call it directly.
+  public func __required<T>() throws -> T where T: Error, Success == T? {
```
Comment (grynspan) [inline Sources/Testing/Support/Additions/ResultAdditions.swift:63]: I am not sure I actually need two separate overloads here.

**Comment by grynspan** [comment]
@swift-ci test

**Comment by grynspan** [comment]
@swift-ci test Linux

**Comment by grynspan** [comment]
@swift-ci test

File: Documentation/Proposals/NNNN-return-errors-from-expect-throws.md [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+## Proposed solution
+
+I propose deprecating [`#expect(_:sourceLocation:performing:throws:)`](https://developer.apple.com/documentation/testing/expect(_:sourcelocation:performing:throws:))
+and [`#require(_:sourceLocation:performing:throws:)`](https://developer.apple.com/documentation/testing/require(_:sourcelocation:performing:throws:))
+and modifying the other overloads so that, on success, they return the errors
+that were thrown.
+
+## Detailed design
+
+All overloads of `#expect(throws:)` and` #require(throws:)` will be updated to
```
Comment (stmontgomery) [inline Documentation/Proposals/NNNN-return-errors-from-expect-throws.md:64]:
```suggestion
All overloads of `#expect(throws:)` and `#require(throws:)` will be updated to
```

File: Documentation/Proposals/NNNN-return-errors-from-expect-throws.md [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+    var potato = Potato()
+    potato.fossilize()
+    try cook(potato) // 🛑 ERROR: Invalid conversion of thrown error type
+                     // 'any Error' to 'PotatoError'
+  }
+  ```
+
+## Acknowledgments
+
+Thanks to the team and to @jakepetroules for starting the discussion that
```
Comment (stmontgomery) [inline Documentation/Proposals/NNNN-return-errors-from-expect-throws.md:263]: FYI this doesn't appear to be a link in the version rendered on GitHub, in case you wanted it to be.

File: Sources/Testing/Testing.docc/AvailabilityStubs/ExpectComplexThrows.md [hunk 6]
```diff
@@ -0,0 +1,28 @@
+# ``expect(_:sourceLocation:performing:throws:)``
+
+<!--
+This source file is part of the Swift.org open source project
+
+Copyright (c) 2023 Apple Inc. and the Swift project authors
```
Comment (stmontgomery) [inline Sources/Testing/Testing.docc/AvailabilityStubs/ExpectComplexThrows.md:6]:
```suggestion
Copyright (c) 2024 Apple Inc. and the Swift project authors
```

File: Sources/Testing/Testing.docc/AvailabilityStubs/RequireComplexThrows.md [hunk 7]
```diff
@@ -0,0 +1,28 @@
+# ``require(_:sourceLocation:performing:throws:)``
+
+<!--
+This source file is part of the Swift.org open source project
+
+Copyright (c) 2023 Apple Inc. and the Swift project authors
```
Comment (stmontgomery) [inline Sources/Testing/Testing.docc/AvailabilityStubs/RequireComplexThrows.md:6]:
```suggestion
Copyright (c) 2024 Apple Inc. and the Swift project authors
```

File: Sources/TestingMacros/Support/Additions/MacroExpansionContextAdditions.swift (hunk 2 above)
Comment (stmontgomery) [reply Sources/TestingMacros/Support/Additions/MacroExpansionContextAdditions.swift:94]: Clever

File: Documentation/Proposals/NNNN-return-errors-from-expect-throws.md (hunk 5 above)
Comment (grynspan) [reply Documentation/Proposals/NNNN-return-errors-from-expect-throws.md:263]: I was expecting it to render as Jake's user ID, but if it doesn't, it's not a big deal. I'll double-check I spelled it correctly either way.

File: Documentation/Proposals/NNNN-return-errors-from-expect-throws.md [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+  thrown, then the test has already failed and is in an inconsistent state, so
+  we should allow the test to fail rather than return what amounts to "bad
+  output".
+  
+  If the test author wants to inspect any arbitrary thrown error, they can
+  specify `(any Error).self` instead of a concrete error type.
+  
+- Adopting [typed throws](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0413-typed-throws.md)
+  to statically require that the error thrown from test code is of the correct
+  type.
```
Comment (stmontgomery) [inline Documentation/Proposals/NNNN-return-errors-from-expect-throws.md:243]: In an offline discussion, @grynspan and I discussed typed throws more deeply and there _may_ be a path to adopting it while avoiding the source compatibility problem described below. That would be a follow-on enhancement to this proposal, but I think we should eventually find a way to support typed throws _for APIs which have adopted it_.

Given that, could we please rephrase this to leave the door open to it, and move it to the "Future directions" section above?

Comment (grynspan) [reply Documentation/Proposals/NNNN-return-errors-from-expect-throws.md:243]: Will do.

File: Documentation/Proposals/NNNN-return-errors-from-expect-throws.md (hunk 5 above)
Comment (grynspan) [reply Documentation/Proposals/NNNN-return-errors-from-expect-throws.md:263]: Added an explicit link.

**Comment by grynspan** [comment]
@swift-ci test

**Comment by grynspan** [comment]
@swift-ci test

**Comment by grynspan** [comment]
@swift-ci test

**Review by stmontgomery — APPROVED** [review]

**Comment by grynspan** [comment]
This is a limitation of Apple's documentation system. See [this](https://forums.swift.org/t/swift-testing-docs-include-unreleased-apis-with-no-indication/78021) forum thread for more discussion.
